drivers/spi/spi_omap_mcspi.c

```diff
@@ -0,0 +1,600 @@
+		/* rx-only */
+		data->chconf |= OMAP_MCSPI_CHCONF_TRM_RX_ONLY;
+
+		/* more than one word */
+		if (ctx->rx_len > 1) {
+			data->chconf |= OMAP_MCSPI_CHCONF_TURBO;
+		}
+	} else if (!rx) {
+		/* tx-only */
+		data->chconf |= OMAP_MCSPI_CHCONF_TRM_TX_ONLY;
+	}
+
+	/* write chconf and chctrl */
+	regs->CHAN[chan].CHCONF = data->chconf;
+	regs->CHAN[chan].CHCTRL = data->chctrl;
+
+	/* enable channel */
+	omap_mcspi_channel_enable(dev, true);
+
+	if (!tx) {
+		/* for rx-only, we must write something to TX */
+		regs->CHAN[chan].TX = 0;
+	}
+
+	/* we only support PIO for now
+	 * TODO: add DMA
+	 */
+	rv = omap_mcspi_transceive_pio(dev, count);
+	if (rv) {
+		return -EIO;
+	}
+
+	/* update rx buffer */
+	spi_context_update_rx(ctx, data->dfs, count);
+
+	/* update tx buffer */
+	spi_context_update_tx(ctx, data->dfs, count);
+
+	return 0;
+}
+
+static int omap_mcspi_transceive_all(const struct device *dev, const struct spi_config *config,
+				     const struct spi_buf_set *tx_bufs,
+				     const struct spi_buf_set *rx_bufs, bool asynchronous,
+				     spi_callback_t callback, void *userdata)
+{
+	struct omap_mcspi_data *data = DEV_DATA(dev);
+	struct spi_context *ctx = &data->ctx;
+	int ret = 0;
+
+	if (!tx_bufs && !rx_bufs) {
+		return 0;
+	}
+
+	spi_context_lock(ctx, asynchronous, callback, userdata, config);
+
+	ret = omap_mcspi_configure(dev, config, tx_bufs, rx_bufs);
+	if (ret) {
+		LOG_ERR("An error occurred in the SPI configuration");
+		goto cleanup;
+	}
+
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, data->dfs);
+
+	spi_context_cs_control(ctx, true);
+
+	while (spi_context_tx_on(ctx) || spi_context_rx_on(ctx)) {
+		ret = omap_mcspi_transceive_one(dev);
+
+		if (ret < 0) {
+			LOG_ERR("Transaction failed, TX/RX left: %d/%d",
+				spi_context_tx_len_left(ctx), spi_context_rx_len_left(ctx));
+			goto cleanup;
+		}
+	}
+
+cleanup:
+	spi_context_cs_control(ctx, false);
+
+	if (!(ctx->config->operation & SPI_LOCK_ON)) {
+		spi_context_release(ctx, ret);
+	}
+	return ret;
+}
+
+static int omap_mcspi_transceive(const struct device *dev, const struct spi_config *config,
+				 const struct spi_buf_set *tx_bufs,
+				 const struct spi_buf_set *rx_bufs)
+{
+	return omap_mcspi_transceive_all(dev, config, tx_bufs, rx_bufs, false, NULL, NULL);
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int omap_mcspi_transceive_async(const struct device *dev, const struct spi_config *config,
+				       const struct spi_buf_set *tx_bufs,
+				       const struct spi_buf_set *rx_bufs, spi_callback_t callback,
+				       void *userdata)
+{
+	/* wait for DMA to be implemented to use IRQ and ASYNC */
+	return -ENOTSUP;
+}
+#endif /* CONFIG_SPI_ASYNC */
+
+static int omap_mcspi_init(const struct device *dev)
+{
+	const struct omap_mcspi_cfg *cfg = DEV_CFG(dev);
+	struct omap_mcspi_data *data = DEV_DATA(dev);
+	struct omap_mcspi_regs *regs = DEV_REGS(dev);
+	volatile uint32_t *sysstatus = &regs->SYSSTATUS;
+	int ret;
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
```

Comment:
`DEV_REGS` should not be used before `DEVICE_MMIO_MAP` is called. This will fail in RAM MMIO.
See https://github.com/zephyrproject-rtos/zephyr/pull/90161 and https://github.com/zephyrproject-rtos/zephyr/pull/89788

Reply:
Oh thanks, completely missed this